main/src/ca/uwaterloo/flix/language/phase/Lexer.scala

```diff
@@ -21,51 +21,670 @@ import ca.uwaterloo.flix.language.ast.{Ast, TokenErrorKind, ReadAst, SourceKind,
+        advance()
+        return TokenKind.HoleVariable
+      } else if (!p.isLetter && !p.isDigit && p != '_' && p != '!') {
+        return kind
+      }
+      advance()
+    }
+    kind
+  }
+
+
+  /**
+   * Moves current position past a java name. IE. "##java"
+   */
+  private def javaName()(implicit s: State): TokenKind = {
+    advance()
+    while (!isAtEnd()) {
+      val p = peek()
+      if (!p.isLetter && !p.isDigit && p != '_' && p != '!') {
+        return TokenKind.NameJava
+      }
+      advance()
+    }
+    TokenKind.NameJava
+  }
+
+  /**
+   * Moves current position past a greek name.
+   * Greek names must lie in the unicode range U+0370 to U+03FF.
+   * IE. "Χαίρετε"
+   */
+  private def greekName()(implicit s: State): TokenKind = {
+    while (!isAtEnd()) {
+      if (!isGreekNameChar(peek())) {
+        return TokenKind.NameGreek
+      }
+      advance()
+    }
+    TokenKind.NameGreek
+  }
+
+  /**
+   * Checks whether `c` lies in unicode range U+0370 to U+03FF
+   */
+  private def isGreekNameChar(c: Char): Boolean = {
+    val i = c.toInt
+    i >= 880 && i <= 1023
+  }
+
+  /**
+   * Moves current position past a math name.
+   * Math names must lie in the unicode range U+2190 to U+22FF
+   * IE. "⊆"
+   */
+  private def mathName()(implicit s: State): TokenKind = {
+    while (!isAtEnd()) {
+      if (!isMathNameChar(peek())) {
+        return TokenKind.NameMath
+      }
+      advance()
+    }
+    TokenKind.NameMath
+  }
+
+  /**
+   * Checks whether `c` lies in unicode range U+2190 to U+22FF
+   */
+  private def isMathNameChar(c: Char): Boolean = {
+    val i = c.toInt
+    i >= 8592 && i <= 8959
+  }
+
+  /**
+   * Moves current position past a named hole. IE. "?foo".
+   */
+  private def namedHole()(implicit s: State): TokenKind = {
+    while (!isAtEnd()) {
+      if (!peek().isLetter) {
+        return TokenKind.HoleNamed
+      }
+      advance()
+    }
+    TokenKind.HoleNamed
+  }
+
+
+  /**
+   * Moves current position past an infix function.
+   */
+  private def infixFunction()(implicit s: State): TokenKind = {
+    while (!isAtEnd()) {
+      if (peek() == '`') {
+        advance()
+        return TokenKind.InfixFunction
+      }
+      advance()
+    }
+    TokenKind.Err(TokenErrorKind.UnterminatedInfixFunction)
+  }
+
+  /**
+   * Moves current position past a user defined operator. IE. "<*>".
+   * A user defined operator may be any combination of length 2 or more
+   * of the characters in `validUserOpTokens`.
+   */
+  private def userDefinedOp()(implicit s: State): TokenKind = {
+    while (!isAtEnd()) {
+      if (!validUserOpTokens.contains(peek())) {
+        return TokenKind.UserDefinedOperator
+      } else {
+        advance()
+      }
+    }
+    TokenKind.UserDefinedOperator
   }
 
-  private def advance()(implicit s: State): Char = ???
+  /**
+   * The characters allowed in a user defined operator mapped to their `TokenKind`s
+   */
+  private val validUserOpTokens = Map(
```

Comment:
This should be next to MAX_BLOCK_COMMENT_NESTING_LEVEL since it is a "static field".